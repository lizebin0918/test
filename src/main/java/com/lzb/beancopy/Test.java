package com.lzb.beancopy;

import com.alibaba.fastjson.JSON;
import lombok.Data;
import org.springframework.cglib.beans.BeanCopier;
import org.springframework.cglib.core.Converter;

import java.time.LocalDate;
import java.util.*;

/**
 * <br/>
 * Created on : 2021-08-26 15:17
 *
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        /*User a = new User();
        a.setAge(18);
        a.setBirthday(LocalDate.now());
        User b = new User();
        b.setAge(1);
        b.setName("lizebin");
        BeanUtils.copyNonNull(a, b);
        System.out.println(JSON.toJSONString(b));

        a = new User();
        a.setBirthday(LocalDate.now());
        b = new User();
        b.setAge(1);
        b.setName("lizebin12344");
        BeanUtils.copyNonNull(a, b);
        System.out.println(JSON.toJSONString(b));

        BeanUtils.copy(a, b);
        System.out.println(JSON.toJSONString(b))*/;

        User a = new User();
        a.setAge(18);
        a.setBirthday(LocalDate.now());
        a.setIntegers(Arrays.asList(1, 2, 3, 4));
        a.setId("1");
        a.setName("name");
        a.setArray(new Integer[]{1, 2});
        UserVo b = new UserVo();
        b.setIsDel(false);

        // BeanUtils.copy(a, b);
        // System.out.println(JSON.toJSONString(a));
        // System.out.println(JSON.toJSONString(b));

        BeanUtils.copyNonEmpty(a, b);
        System.out.println(JSON.toJSONString(b));

        System.out.println(new ArrayList<>() instanceof RandomAccess);

        System.out.println(ArrayList.class.isNestmateOf(List.class));
        System.out.println(List.class.isAssignableFrom(ArrayList.class));
        System.out.println(ArrayList.class.isAssignableFrom(ArrayList.class));

        /*var a1 = Arrays.asList(a, b);
        var newList = BeanUtils.copyList(a1, () -> new UserVo());
        System.out.println(JSON.toJSONString(newList));*/

        /*Integer x = 3;
        Integer y = 4;
        BeanUtils.copy(x, y);
        System.out.println(y);

        Integer i = new Integer(100);
        org.springframework.beans.BeanUtils.copyProperties(new Integer(1), i);
        System.out.println(i);*/

        // 给定一个值，对数组元素扣减，扣完为零
       /* int count = 3;
        List<Integer> list = Arrays.asList(1, 0, 0, 1, 1, 0, 0);
        for (int i = 0; i < list.size(); i++) {
            int item = list.get(i);
            if (item <= count) {
                count = count - item;
                list.set(i, count);
            } else {
                list.set(i, item - count);
                count = 0;
            }
            if (count == 0) {
                break;
            }
        }
        System.out.println(list);
        System.out.println(count);*/

    }

    @Data
    public static class UserVo {
        private Integer id;
        private String name;
        private LocalDate birthday;
        private ArrayList<Integer> integers;
        private Boolean isDel;
        private Integer[] array;
    }

    @Data
    public static class User {
        private String id;
        private Integer age;
        private String name;
        private LocalDate birthday;
        private List<Integer> integers;
        private Boolean isDel;
        private Integer[] array;
    }

    public static class MyConvertor implements Converter {
        BeanCopier beanCopier;
        public MyConvertor(BeanCopier beanCopier) {
            this.beanCopier = beanCopier;
        }

        /**
         * @param value 源对象值
         * @param targetClass Class:目标对象属性类
         * @param setMethod String:目标对象setter方法名
         * @return
         */
        @Override
        public Object convert(Object value, Class paramClass, Object setMethod) {
            System.out.println("-------------");
            System.out.println(value);
            System.out.println(paramClass);
            System.out.println(setMethod);
            System.out.println("-------------");
            System.out.println(beanCopier.getClass());
            if (Objects.isNull(value)) {
            }
            return value;
        }
    }
}
