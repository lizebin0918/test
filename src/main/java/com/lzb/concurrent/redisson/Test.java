package com.lzb.concurrent.redisson;

import org.redisson.Redisson;
import org.redisson.api.*;
import org.redisson.client.codec.StringCodec;
import org.redisson.config.Config;

import java.util.concurrent.CompletableFuture;

/**
 * 分布式公平锁测试<br/>
 * Created on : 2021-06-03 21:10
 * @author lizebin
 */
public class Test {

    private static final RedissonClient redisson;
    static {
        Config config = new Config();
        config.setCodec(StringCodec.INSTANCE)
            .useSingleServer().setAddress("redis://192.168.56.100:6379")
            .setSslEnableEndpointIdentification(false);
        redisson = Redisson.create(config);
    }

    public static void main(String[] args) throws InterruptedException {
        System.out.println("-------------------");

        RBucket<String> key = redisson.getBucket("1", StringCodec.INSTANCE);
        System.out.println(key.get());

        //Note:这个key不能跟其他键冲突
        RLock lock = redisson.getLock("lock");
        for (int i = 0; i < 10; i++) {
            CompletableFuture.runAsync(() -> {
                String tn = Thread.currentThread().getName();
                lock.lock();
                System.out.println(tn + " get lock ");
                try {
                    System.out.println(tn + " start sleep");
                    Thread.sleep((long) (Math.random() * 500L));
                    System.out.println(tn + " start end");
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } finally {
                    lock.unlock();
                }
            });
        }

        Thread.sleep(100000000);
    }

}
