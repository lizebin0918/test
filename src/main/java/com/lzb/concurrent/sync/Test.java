package com.lzb.concurrent.sync;

import java.io.IOException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * 只用synchronized实现生产者消费者模式：A线程打印奇数，B线程打印偶数<br/>
 * Created on : 2021-05-19 09:24
 * @author chenpi 
 */
public class Test {

    public static void main(String[] args) throws IOException, InterruptedException {

        System.out.println(3 & 0);

        AtomicInteger count = new AtomicInteger();
        Object lock = new Object();
        int max = 100;
        Thread a = new Thread(() -> {
            while (count.intValue() < max) {
                synchronized (lock) {
                    if ((count.intValue() & 1) == 0) {
                        System.out.println("a----->" + count.getAndIncrement());
                    }
                }
            }
        });
        Thread b = new Thread(() -> {
            while (count.intValue() < max) {
                synchronized (lock) {
                    if ((count.intValue() & 1) == 1) {
                        System.out.println("b----->" + count.getAndIncrement());
                    }
                }
            }
        });
        a.start();
        b.start();
        a.join();
        b.join();

        System.out.println("done");
    }


}
