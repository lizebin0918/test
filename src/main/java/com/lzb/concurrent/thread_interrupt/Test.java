package com.lzb.concurrent.thread_interrupt;
/**
 * sleep 被中断，抛异常，会自动重置中断标志位，isInterrupted 是 false
 * 如果要中断线程，可以调用：Thread.currentThread().interrupt()；
 * 如果要返回当前中断标志位，并复位：Thread.interrupted()
 * 如果线程休眠，关闭应用：Process finished with exit code 130 (interrupted by signal 2: SIGINT)
 *
 * Created on : 2021-05-17 10:49
 * @author chenpi 
 */
public class Test {

    public static void main(String[] args) {
        Thread t = new Thread(() -> {
            while (true) {
                if (Thread.currentThread().isInterrupted()) {
                    System.out.println("线程被中断了，退出");
                    break;
                }
                System.out.println("1");
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    try {
                        Thread.sleep(100000);
                    } catch (InterruptedException e1) {
                        e1.printStackTrace();
                    }
                    e.printStackTrace();
                    System.out.println("sleep 被中断");
                    //Thread.currentThread().interrupt();
                }
            }
        });
        t.start();
        try {
            Thread.sleep(50000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        //t.interrupt();
    }

}
