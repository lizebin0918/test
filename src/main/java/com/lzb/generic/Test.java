package com.lzb.generic;

import com.alibaba.fastjson.JSON;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.val;

import java.util.ArrayList;
import java.util.List;

/**
 * 泛型的设计：为了保证类型的读写安全<br/>
 * Created on : 2022-03-12 20:19
 *
 * @author cidervisioncase
 */
public class Test {

    public static void main(String[] args) {

        Test client = new Test();

        Apple apple = new Apple();
        apple.name = "apple";

        // Food
        List<Food> foods = new ArrayList<>();
        foods.add(apple);
        //client.extendsFruit(foods); 编译错误，类型非法，实物不是水果的子类
        client.superFruit(foods);

        // Fruit
        List<Fruit> fruits = new ArrayList<>();
        fruits.add(apple);
        client.superFruit(fruits);
        client.extendsFruit(fruits);

        // Apple
        List<Apple> apples = new ArrayList<>();
        apples.add(apple);
        // client.superFruit(apples);编译错误，类型非法，苹果不是水果的超类
        client.extendsFruit(apples);

    }

    @Data
    private static class Food {
        protected String name;
    }

    @Data
    @EqualsAndHashCode(callSuper = true)
    private static class Fruit extends Food {

    }

    @Data
    @EqualsAndHashCode(callSuper = true)
    private static class Apple extends Fruit {

    }

    public void extendsFruit(List<? extends Fruit> extendsFruits) {

        System.out.println("水果子类列表:" + JSON.toJSONString(extendsFruits));

        //extendsFruits.add(XXXXX); 编译异常，这样就可以添加：苹果、香蕉、梨，违反了泛型的初衷

        // 编译通过：因为集合元素都是水果的子类
        Fruit apple = extendsFruits.get(0);
    }

    public void superFruit(List<? super Fruit> superFruits) {

        System.out.println("水果超类列表:" + JSON.toJSONString(superFruits));

        // 因为容器的类型是Fruit的父类，那就是说容器可以放：水果及水果的子类，这样也是类型安全的
        //superFruits.add(new Food());编译报错
        superFruits.add(new Fruit());
        superFruits.add(new Apple());

        // get 就会报错，因为超类也有很多，get的类型不安全
        //Fruit food = superFruits.get(0);

    }

}
