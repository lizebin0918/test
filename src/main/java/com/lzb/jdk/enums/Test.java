package com.lzb.jdk.enums;

/**
 * <br/>
 * Created on : 2022-03-13 18:23
 *
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        boolean valueExist = EnumUtils.getByValue(TypeEnum.values(), 1).isPresent();
        System.out.println(valueExist);
        System.out.println(EnumUtils.getByValue(TypeEnum.values(), 4).isPresent());
        System.out.println(EnumUtils.getByValue(TypeEnum.class, 4).isPresent());

        System.out.println(EnumUtils.getByValue(StatusEnum.class, "all", String.CASE_INSENSITIVE_ORDER).isPresent());
        System.out.println(EnumUtils.getByValue(StatusEnum.class, "all1", String.CASE_INSENSITIVE_ORDER).isPresent());

    }

}
