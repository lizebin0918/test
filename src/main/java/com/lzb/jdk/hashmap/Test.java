package com.lzb.jdk.hashmap;

import org.apache.commons.lang3.RandomStringUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * <br/>
 * Created on : 2021-10-06 23:56
 *
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        // 扰动函数真的可以让散列更加均匀？
        Map<Integer, Integer> map = new HashMap<>(16);
        int size = 10000;
        List<String> words = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            words.add(RandomStringUtils.random(10));
        }

        for (String word : words) {
            int idx = Disturb.disturbHashIdx(word, 128);
            // int idx = Disturb.hashIdx(word, 128);
            if (map.containsKey(idx)) {
                Integer i = map.get(idx);
                map.put(idx, ++i);
            } else {
                map.put(idx, 1);
            }
        }

        System.out.println(map.values());
    }

    private static class Disturb {

        public static int disturbHashIdx(String key, int size) {
            return (size - 1) & (key.hashCode() ^ (key.hashCode() >>> 16));
        }

        public static int hashIdx(String key, int size) {
            return (size - 1) & key.hashCode();
        }

    }
}
