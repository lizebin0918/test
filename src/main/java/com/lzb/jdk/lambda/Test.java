package com.lzb.jdk.lambda;

import com.alibaba.fastjson.JSON;

import javax.annotation.Nonnull;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * 测试lambda表达式<br/>
 * Created on : 2021-08-21 16:46
 *
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        List<String> sampleList = Arrays.asList("Java", "Kotlin", null);
        String resultString = sampleList.stream()
                .map((@Nonnull var x) -> x.toUpperCase())
                .collect(Collectors.joining(", "));

        System.out.println(sampleList);

        a(new Object());
        a(null);

        Optional.ofNullable(1).orElseThrow(() -> new RuntimeException());
    }

    public static void a(@Nonnull Object o) {
        System.out.println(JSON.toJSONString(o));
    }

}
