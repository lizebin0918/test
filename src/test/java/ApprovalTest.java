import java.io.ByteArrayOutputStream;
import java.util.Arrays;
import java.util.Map;

import com.alibaba.fastjson.JSON;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.approvaltests.ApprovalUtilities;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

/**
 * <br/>
 * Created on : 2023-01-08 09:49
 * @author mac
 */
public class ApprovalTest {

	@Test
	void testList() {
		String[] names = {"Llewellyn", "James", "Dan", "Jason", "Katrina"};
		Arrays.sort(names);
		Approvals.verifyAll("", names);
	}

	@Test
	void should_test_map() {
		Map<String, String> map = Map.of("a", "a", "b", "b");
		Approvals.verify(map);
	}

	/**
	 * 直接toString()
	 */
	@Test
	void should_test_person() {
		Person p = new Person();
		p.setAge(10);
		p.setName("naem");
		Approvals.verify(p);
	}

	@Test
	void test_json() {
		String[] names = {"Llewellyn", "James", "Dan", "Jason", "Katrina"};
		Approvals.verify(JSON.toJSONString(names));
	}

	@Test
	void should_test_sysout_inline_method() {
		ByteArrayOutputStream output = new ApprovalUtilities().writeSystemOutToStringBuffer();

		PersonService personService = new PersonService();
		personService.test(new Person(1, "name"));

		Approvals.verify(SystemOutUtils.convert(output));
	}

	@Data
	@AllArgsConstructor
	@NoArgsConstructor
	static class Person {
		private int age;
		private String name;
	}

	@Slf4j
	static class PersonService {

		public void test(Person person) {
			System.out.println("this is a person age : " + person.getAge());
		}

	}

}
