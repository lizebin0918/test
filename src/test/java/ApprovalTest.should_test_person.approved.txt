ApprovalTest.Person(age=10, name=naem)